internal/resource/config/datastore/custom_data_store.go

```diff
@@ -0,0 +1,222 @@
+	state.LdapDataStore = ldapDataStoreEmptyStateObj
+	state.PingOneLdapGatewayDataStore = pingOneLdapGatewayDataStoreEmptyStateObj
+	return diags
+}
+
+func addOptionalCustomDataStoreFields(addRequest client.DataStoreAggregation, con context.Context, createJdbcDataStore client.CustomDataStore, plan dataStoreResourceModel) error {
+	customDataStorePlan := plan.CustomDataStore.Attributes()
+
+	if internaltypes.IsDefined(plan.CustomId) {
+		addRequest.CustomDataStore.Id = plan.CustomId.ValueStringPointer()
+	}
+
+	parentRef := customDataStorePlan["parent_ref"]
+	if internaltypes.IsNonEmptyObj(parentRef.(types.Object)) {
+		parentRef, err := resourcelink.ClientStruct(parentRef.(types.Object))
+		if err != nil {
+			return err
+		}
+		addRequest.CustomDataStore.ParentRef = parentRef
+	}
+	return nil
+}
+
+func createCustomDataStore(plan dataStoreResourceModel, con context.Context, req resource.CreateRequest, resp *resource.CreateResponse, dsr *dataStoreResource) {
+	var diags diag.Diagnostics
+	var err error
+
+	customPlan := plan.CustomDataStore.Attributes()
+	name := customPlan["name"].(types.String).ValueString()
+	pluginDescriptorRef, err := resourcelink.ClientStruct(customPlan["plugin_descriptor_ref"].(types.Object))
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to create plugin descriptor reference object for DataStore", err.Error())
+		return
+	}
+
+	configuration := &client.PluginConfiguration{}
+	err = json.Unmarshal([]byte(internaljson.FromValue(customPlan["configuration"].(types.Object), true)), configuration)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to create configuration object for DataStore", err.Error())
+		return
+	}
+
+	createCustomDataStore := client.CustomDataStoreAsDataStoreAggregation(client.NewCustomDataStore("CUSTOM", name, *pluginDescriptorRef, *configuration))
+	err = addOptionalCustomDataStoreFields(createCustomDataStore, con, client.CustomDataStore{}, plan)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to add optional properties to add request for DataStore", err.Error())
+		return
+	}
+
+	apiCreateDataStore := dsr.apiClient.DataStoresAPI.CreateDataStore(config.ProviderBasicAuthContext(con, dsr.providerConfig))
+	apiCreateDataStore = apiCreateDataStore.Body(createCustomDataStore)
+	customDataStoreResponse, httpResp, err := dsr.apiClient.DataStoresAPI.CreateDataStoreExecute(apiCreateDataStore)
+	if err != nil {
+		config.ReportHttpError(con, &resp.Diagnostics, "An error occurred while creating the DataStore", err, httpResp)
+		return
+	}
+
+	// Read the response into the state
+	var state dataStoreResourceModel
+	diags = readCustomDataStoreResponse(con, customDataStoreResponse, &state, &plan.CustomDataStore)
+	resp.Diagnostics.Append(diags...)
+	diags = resp.State.Set(con, state)
+	resp.Diagnostics.Append(diags...)
+}
+
+func updateCustomDataStore(plan dataStoreResourceModel, con context.Context, req resource.UpdateRequest, resp *resource.UpdateResponse, dsr *dataStoreResource) {
```

Comment:
These create and update methods seem to have a lot of duplication across the data store types. You could probably move a lot of the common stuff into the main resource file, and have these methods just help with building the relevant part of the plan struct.

Reply:
I've added `createDataStore` and `updateDataStore` methods to the main file for this. 

https://github.com/pingidentity/terraform-provider-pingfederate/blob/052b66d82a2678c0c8bde6766061521a4d904a99/internal/resource/config/datastore/data_store_resource.go#L83

https://github.com/pingidentity/terraform-provider-pingfederate/blob/052b66d82a2678c0c8bde6766061521a4d904a99/internal/resource/config/datastore/data_store_resource.go#L159